code/renderers/react/src/__test-dts__/CSF3.test-d.tsx

```diff
@@ -0,0 +1,148 @@
+    });
+    const Basic: StoryObj<typeof meta> = {
+      args: { disabled: false },
+    };
+  });
+
+  test('❌ The combined shape of meta args and story args must match the required args.', () => {
+    {
+      const meta = satisfies<Meta<typeof Button>>()({ component: Button });
+      const Basic: StoryObj<typeof meta> = {
+        // @ts-expect-error disabled not provided ❌
+        args: { label: 'good' },
+      };
+    }
+    {
+      const meta = satisfies<Meta<typeof Button>>()({
+        component: Button,
+        args: { label: 'good' },
+      });
+      // @ts-expect-error disabled not provided ❌
+      const Basic: StoryObj<typeof meta> = {};
+    }
+    {
+      const meta = satisfies<Meta<ButtonProps>>()({ component: Button });
+      const Basic: StoryObj<typeof meta> = {
+        // @ts-expect-error disabled not provided ❌
+        args: { label: 'good' },
+      };
+    }
+  });
+});
+
+test('✅ All void functions are optional', () => {
+  interface CmpProps {
+    label: string;
+    disabled: boolean;
+    onClick(): void;
+    onKeyDown: KeyboardEventHandler;
+    onLoading: (s: string) => JSX.Element;
+    submitAction(): void;
+  }
+
+  const Cmp: (props: CmpProps) => JSX.Element = () => <></>;
+
+  const meta = satisfies<Meta<CmpProps>>()({
+    component: Cmp,
+    args: { label: 'good' },
+  });
+
+  const Basic: StoryObj<typeof meta> = {
+    args: { disabled: false, onLoading: () => <div>Loading...</div> },
+  };
+});
+
+type ThemeData = 'light' | 'dark';
+declare const Theme: (props: { theme: ThemeData; children?: ReactNode }) => JSX.Element;
+
+describe('Story args can be inferred', () => {
+  test('Correct args are inferred when type is widened for render function', () => {
+    type Props = ButtonProps & { theme: ThemeData };
+
+    const meta = satisfies<Meta<Props>>()({
+      component: Button,
+      args: { disabled: false },
+      render: (args, { component }) => {
+        // component is not null as it is provided in meta
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const Component = component!;
+        return (
+          <Theme theme={args.theme}>
+            <Component {...args} />
+          </Theme>
+        );
+      },
+    });
+
+    const Basic: StoryObj<typeof meta> = { args: { theme: 'light', label: 'good' } };
+  });
+
+  const withDecorator: DecoratorFn<{ decoratorArg: number }> = (Story, { args }) => (
+    <>
+      Decorator: {args.decoratorArg}
+      <Story args={{ decoratorArg: 0 }} />
+    </>
+  );
+
+  test('Correct args are inferred when type is widened for decorators', () => {
+    type Props = ButtonProps & { decoratorArg: number };
+
+    const meta = satisfies<Meta<Props>>()({
+      component: Button,
+      args: { disabled: false },
+      decorators: [withDecorator],
+    });
+
+    // Yes, decorator arg is required
+    const Basic: StoryObj<typeof meta> = { args: { decoratorArg: 0, label: 'good' } };
+  });
```

Comment:
I'm interested in what happens when you have > 1 decorator (or loader, say). Is this line:

```
decorators?: DecoratorFunction<ReactFramework, infer DArgs>[];
```

Going to infer the right thing for `DArgs` in that case? Maybe a test to confirm it would be good.

Reply:
Will do!

Reply:
This is actually a problem, good catch!
There is a workaround, to basically, safely upcast it...
` decorators: [withDecorator, secondDecorator] as DecoratorFn<Props>[],`
or a factory function of course ...
`decorators: defineDecorators([withDecorator, secondDecorator]),`
Gonna play around if some more magic could do the trick... I kind of hoped the infer keyword to do a bit more here...

A even bigger problem is that if the inferring fails, TS kind of falls into the other branch, which is strange, because it really is a DecoratorFunction...

Reply:
Found a solution, by doing it in two steps:

```ts

type DecoratorsArgs<Decorators> = UnionToIntersection<
  Decorators extends DecoratorFn<infer Args1> ? Args1 : unknown
>;

export type StoryObj<MetaOrArgs = Args> = MetaOrArgs extends {
  render?: ArgsStoryFn<ReactFramework, infer RArgs>;
  component?: ComponentType<infer CmpArgs>;
  loaders?: LoaderFunction<ReactFramework, infer LArgs>[];
  args?: infer DefaultArgs;
  decorators?: (infer Decorators)[];
}
  ? CmpArgs & RArgs & DecoratorsArgs<Decorators> & LArgs extends infer TArgs
```

Same can be done for loaders...

Phew 😅

Reply:
🙌